retrieval/train.py

```diff
@@ -0,0 +1,104 @@
+# Train a retrieval transformer with a dataset
+
+import torch
+import rich
+from rich.console import Console
+from rich.progress import (
+    Progress,
+    BarColumn,
+    SpinnerColumn,
+    TimeElapsedColumn,
+    TextColumn,
+)
+
+
+from retro_pytorch import RETRO, TrainingWrapper
+
+console = Console()
+
+# instantiate RETRO, fit it into the TrainingWrapper with correct settings
+
+retro = RETRO(
+    max_seq_len=2048,
+    enc_dim=896,
+    enc_depth=3,
+    dec_dim=768,
+    dec_depth=12,
+    dec_cross_attn_layers=(1, 3, 6, 9),
+    heads=8,
+    dim_head=64,
+    dec_attn_dropout=0.25,
+    dec_ff_dropout=0.25,
+).cuda()
+
+# mkdir ./chunks - or this will fail
+# data set must be *.md files in /home/sdake/en
+# I used: https://github.com/istio/istio.io/tree/master/content/en
+wrapper = TrainingWrapper(
+    retro=retro,
+    knn=2,
+    chunk_size=64,
+    documents_path="/home/sdake/en",
+    glob="**/*.md",
+    chunks_memmap_path="./chunks/train.chunks.dat",
+    seqs_memmap_path="./chunks/train.seq.dat",
+    doc_ids_memmap_path="./chunks/train.doc_ids.dat",
+    max_chunks=1_000_000,
+    max_seqs=100_000,
+    knn_extra_neighbors=100,
+    max_index_memory_usage="100G",
+    current_memory_available="100G",
+)
+
+retro.train()
+optim = wrapper.get_optimizer(lr=3e-4, wd=0.01)
+
+from rich.progress import (
+    BarColumn,
+    MofNCompleteColumn,
+    Progress,
+    TextColumn,
+    TimeElapsedColumn,
+    TimeRemainingColumn,
+)
+
+#    TextColumn("loss: {}".format(disp_loss)), - this text column needs to be modifiable
+disp_loss = 0.0
+# Define custom progress bar
+progress_bar = Progress(
+    TextColumn("[progress.description]{task.description}"),
+    TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
+    SpinnerColumn(),
+    BarColumn(),
+    #    MofNCompleteColumn(),
+    TextColumn("•"),
+    TimeElapsedColumn(),
+    TextColumn("•"),
+    TimeRemainingColumn(),
+)
+
+
+console.print(
+    "[link=https://github.com/artificialwisdomai/origin]Artificial Wisdom™[/] Retreival Transformer Training",
+    style="#008080",
+)
+console.print(
+    "• retrieval_model=artificialwisdomai/retroformer • foundation_model=mosaicml/mpt30b •",
+    style="#008080",
+)
+
+epoch_max = 15
```

Comment:
Unused constant.

Reply:
Added issue #54 to address this.

Reply:
This is resolved in my updated PR. I am not sure it renders the tasks correctly. For now, I will resolve this conversation.